Maintain AR connection pools in the background

matthewd opened the pull request:
Recruit the Reaper to do more work in keeping the connection pool healthy:

* retiring connections that have been idle[^1] for too long, or that have been open beyond a configured total maximum age
* creating occasional activity on inactive[^1] connections
* keeping the pool prepopulated up to its minimum size
* proactively connecting to the target database from any pooled connections that had lazily deferred that step
* resetting or replacing connections that are known to be broken


For most applications this is going to make no difference. For large applications with multiple active connection pools, though, it will allow more maintenance work to be done in the background and/or in parallel. 


Maximum total connection lifetime is helpful, for example, to facilitate smooth database server failover: as long as the old server remains available for (max lifetime) seconds after a load balancer switches to a new server, no in-flight activity will be interrupted.


Subsumes and thus closes #50990; fixes #50989


[^1]: "idle" and "inactive" here distinguish between connections that have not been requested by the application in a while (idle) and those that have not spoken to their remote server in a while (inactive). The former is a desirable opportunity to reduce our connection count (`idle_timeout`); the latter is a risk that the server or a firewall may drop a connection we still anticipate using (avoided by `keepalive`).

matthewd committed:
Move raw-connection-mangling methods up to adapter helpers

We should still avoid using them whenever possible, but e.g. I'd like to
be able to break connections inside a pool test.

matthewd committed:
Override isolation level before common test setup

The shared setup method creates a connection pool, and having the
isolation level change after a pool is created makes things awkward.

This has slightly more duplication, but not enough that I feel the need
to seek a different abstraction.

matthewd committed:
Allow connections to be borrowed out of the pool for maintenance

matthewd committed:
Simplify the construction of once-off pool configurations

In the process, be more consistent about always disconnecting pools
we've created.

matthewd committed:
Allow a connection pool to iteratively maintain its connections

Building upon the "checkout for maintenance" primitives, this gives us a
method to progressively loop over idle connections without the risk of
starving the pool while we work.

Even for something internal, I don't love the API... but hopefully it
will suffice until we come up with something better.

matthewd committed:
Maintain a minimum pool size when expiring idle connections

Co-authored-by: Chris AtLee <chris.atlee@shopify.com>

matthewd committed:
Proactively populate the pool up to minimum after it is activated

Co-authored-by: Chris AtLee <chris.atlee@shopify.com>

matthewd committed:
Automatically activate pool when a checkout occurs in a second context

matthewd committed:
Preconnect connections that are idle in the pool

matthewd committed:
Application-layer database keepalive

Optionally ensure the full database connection chain sees regular query
traffic (without affecting our internal idle counters).

matthewd committed:
Maximum age for database connections

Avoid using database connections that were originally established over
the configured duration ago. This can be helpful to provide smooth
failover between connection proxies.

matthewd committed:
Allow the entire connection pool to be recycled on demand

This is ideal where the caller knows that a database server/proxy
failover is occurring (meaning that previously-established connections
are now potentially pointing to the wrong place, and new fresh
connections will go to the right one); where such signalling is
possible, it can be used in place of an onerously low / "preventative"
max_age.

matthewd committed:
Resolve the ref once at the start of the block

It's less elegant, but just seems easier to reason about.

matthewd committed:
Update Reaper to reflect its broader responsibilities

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -727,11 +845,109 @@ def build_async_executor
+        # before the first call has completed. (Though regular pool behaviour
+        # will prevent two instances from working on the same specific
+        # connection at the same time.)
+        def sequential_maintenance(candidate_selector, &maintenance_work)
```
I don't like this. Am I missing some much simpler way to achieve a similar result?

byroot commented:
It would also be desirable to trigger this code path before getting ready to serve traffic, generally after fork but depends.

ChaelCodes committed:
Deprecate pool config value in favor of max_connections and min_connections

Co-authored-by: Matthew Draper <matthew@trebex.net>

ChaelCodes committed:
Add changelog entry for new pool configs

Co-authored-by: Matthew Draper <matthew@trebex.net>

nateberkopec commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Introduce new database configuration options `keep_alive`, `max_age`, and
```
I think you mean `keepalive`

nateberkopec replied:
nit: tbh I understand why you didn't snake_case it due to precedent on how we use this word in network contexts, but what the setting _does_ here isn't the same as that context, so I think it should be snakecased.

andyatkinson replied:
From a developer usability perspective it would be nice to use one form or the other consistently.

andyatkinson commented:
> creating occasional activity on inactive1 connections

Hello. I was curious if there is an example of what this will enable that wasn't possible before.

I also read the footnote but don't grasp the distinction between idle and inactive.

In Postgres a connection may only be "idle" (inside or outside a transaction) besides being "active". The binary states have a certain elegance in simplicity.

We can (and usually should) set a timeout for a transaction that's not doing work (idle_in_transaction_session_timeout), but there's not a concept of a "inactive" connection state. Wondering if SQLite or MySQL/MariaDB are different.

Thanks for help in understanding the use cases for this being proposed.

matthewd commented:
I'm not sure how much more context I can offer beyond what's in the footnote: keepalive will ensure the backend server (or proxy) sees occasional query traffic, so it will know we're still present. In Postgres terms, it complements a server's `idle_session_timeout` -- which has no corresponding connection state, because when it's passed, the connection gets closed.

These are settings for configuring how the pool manages its set of connections, rather than backend-specific behaviour around individual connections.

I'd say all the new options are going to fall into the category of "if you don't know that you need this [and thus go looking for it], you don't need to consider/set it". 

> In Postgres a connection may only be "idle" (inside or outside a transaction) besides being "active". The binary states have a certain elegance in simplicity.

Are you talking about `PQtransactionStatus`? That's not really a general connection state... and is also not binary, even within its constrained definition. I love a good oversimplification, but I don't know that it creates a strong base to then compare on a simplicity scale.

mfournier commented:
Wow thanks for this work @matthewd :muscle: this will be super useful for us. We have half a dozen workarounds and bandaids in place for the class of problems this solves !

@andyatkinson the purpose (in my understanding) is rather to offer better handling of the TCP connections from Rails to the database. I totally agree idle transactions and such should be timeouted, but this will only release a slot back to the connection pool. The TCP session underneath (+ TLS & authentication) will be reused later.

Generally speaking networking tools (firewalls, load-balancers, service-mesh, etc) frown upon long-lived idle connections: it uses up ressources, no way to know if they're zombies or not. So in practice sooner or later a timeout kicks in. Then when suddenly Rails get a surge of traffic, these dormant pool slots get woken up, discover that the other end is no longer responding, and raise an `ActiveRecord::ConnectionNotEstablished`, generally after some painfully long milliseconds of time.

Another use case is when you add a reader to a cluster behind a load-balancer, you have to restart your whole fleet of Rails instances to rebalance the DB connections evenly between all the members, because they have no reason to reopen existing connections in the pool, and no way to know the topology has changed. Same thing the other way round: when you need to drain existing connections to a DB server before retiring it, without breaking in-flight DB queries.

HTH understanding the purpose 🙂
And thanks again Matthew, I hope your PR will get included in Rails, it would be a great addition !

matthewd committed:
Apply a pool_jitter adjustment to keepalive and max_age

For each connection, we'll reduce both values by up to 20% (by default),
preventing repeated thundering herds.

Co-authored-by: ChaelCodes <chaelcodes@gmail.com>

matthewd committed:
Reduce default reaping frequency now that it has more jobs

Lower the baseline to 20s, but also reduce it further if that's needed
in order to usefully achieve one of our configured timeouts.

Co-authored-by: ChaelCodes <chaelcodes@gmail.com>

andyatkinson commented:
Hi @matthewd and @mfournier, I chimed in on this PR while on vacation, prompted by social media, probably should’ve waited. 😁

To clarify my intent: I was advocating against introducing another connection concept like "inactive" or connection retirement if possible. Users (devs and operations) already face challenges managing DB connections efficiently across the server, client, and poolers, even with the existing, more minimal set of states, so my general mindset for connection management features is "less is more."

Understanding resource consumption with an existing set of connections is challenging in my opinion, although admittedly beyond the scope of Rails. Using that info to size the number of connections appropriately to avoid overspending or under allocating is difficult. I've worked as an employee and as a database consultant for orgs that want to better understand their usage and then size levels appropriately.

Connection poolers like PgBouncer introduce another dimension of complexity.

I’m not super familiar with libpq functions like PQtransactionStatus, but I see it provides detailed granularity, including multiple idle states within a transaction. From that perspective, I agree it's not really "binary" as in a connection is either active or idle.
https://www.postgresql.org/docs/17/libpq-status.html

It sounds like the capabilities in this PR are going to be helpful for some orgs.

Thus my mindset is to shift focus more to seeing if I can offer help on the API docs, Rails Guides, or by blogging etc, to help users determine if it will be helpful to them.

Thanks!

andyatkinson commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Introduce new database configuration options `keepalive`, `max_age`, and
+    `min_connections` -- and rename `pool` to `max_connections` to match.
```
I like the `min_connections` and `max_connections` rename to help clarify that "max connections" means "up to" the number would be in use when activity is high.

andyatkinson commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool/reaper.rb:
```diff
@@ -7,12 +7,29 @@ module ConnectionAdapters
+      #  * retiring connections that have been idle[1] for too long
+      #  * creating occasional activity on inactive[1] connections
+      #  * keeping the pool prepopulated up to its minimum size
+      #  * proactively connecting to the target database from any pooled
```
What about "eagerly" instead of "proactively", for more consistency with eager load/lazy loading from elsewhere in Rails that some users will already be familiar with? Just an idea.

andyatkinson commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool/reaper.rb:
```diff
@@ -7,12 +7,29 @@ module ConnectionAdapters
+      #
+      # Beyond that, it manages the health of available / unleased connections:
+      #  * retiring connections that have been idle[1] for too long
+      #  * creating occasional activity on inactive[1] connections
```
For "creating occasional activity" is it worth mentioning the "keepalive" here?

andyatkinson commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool/reaper.rb:
```diff
@@ -7,12 +7,29 @@ module ConnectionAdapters
-      # +pool+. A reaper instantiated with a zero frequency will never reap
-      # the connection pool.
+      # The reaper is a singleton that exists in the background of the process
+      # and is responsible for general maintenance of all the connection pools.
```
> is responsible for general maintenance of all the connection pools.

There's only one pool right?

If so, could we say "maintenance of the connections"? This is subtle, but I think using a singular pool term here helps because we do also run multiple instances of an application where each has its own pool, and within a pool we expect to have multiple connections.

benoittgt replied:
Maybe it's relevant with `multi_thread_pool` with [async queries](https://guides.rubyonrails.org/configuring.html#config-active-record-async-query-executor) ?

ChaelCodes committed:
Test with larger idle timeouts when we're not sleeping

matthewd committed:
Remove a false claim from the documentation

The reaper thread has long been responsible for collecting dead threads'
connections; that doesn't happen in `clear_active_connections!`.

matthewd committed:
Unify definitions of with_temporary_connection_pool

Also, add a couple of disconnect! calls for other locally-constructed
temporary pools.

matthewd committed:
No need to lease a connection just to check adapter type

matthewd committed:
Keep a central definition of whether there's any reaping work to do

matthewd committed:
Lock the pool while performing global operations

If a caller wants to do something that affects the entire pool contents
(like full #disconnect), we need to ensure we're not fighting with the
reaper.

matthewd committed:
More checks for discarded pool

matthewd committed:
After full disconnect, stop maintaining minimum size

It's unlikely the caller expects the pool to immediately repopulate.

matthewd committed:
Be more structured in selecting pools for reaping

adis-io commented:
> Another use case is when you add a reader to a cluster behind a load-balancer, you have to restart your whole fleet of Rails instances to rebalance the DB connections evenly between all the members, because they have no reason to reopen existing connections in the pool, and no way to know the topology has changed. Same thing the other way round: when you need to drain existing connections to a DB server before retiring it, without breaking in-flight DB queries.

Having this issue atm. We have pgcat as a DB pooler and when we need to retire one of deployments – we can't, because rails uses the connection indefinitely. You mentioned bandaids, could you please share what kind of is that? We are still on 6.0 so even if this PR gets merged, we won't be able to use it unless someone backports it to previous versions

fractaledmind commented:
I just want to add that adding `min_connections` will make SQLite-backed apps better, as it will remove the bottleneck of write throughput saturating the connection pool and holding back reads, even tho at the db level reads can always occur. This one feature will be a notable performance improvement for SQLite apps with bursts of heavy write traffic.

matthewd committed:
Don't deprecate the 'pool' option yet

We'll prefer the new 'max_connections' name for documentation etc, but
unless the user is introducing a confusing situation by adding
'min_connections', we don't need to force config churn during upgrade.

ngan commented:
Hi 👋 @matthewd - how's it going? Haven't seen you in awhile, hope you're doing well! 

At Gusto, we tried upgrading to Rails 8.1 but ran into an issue where our requests just sat there a good 85s doing nothing. After some debugging, I _think_ the problem is with this particular change. Here's the details:

We hit a deadlock in this PR's `try_to_queue_for_background_connection` that caused ~`85`-second request blocks in production. `try_to_queue_for_background_connection` holds the pool's `MonitorMixin` while waiting on the queue's `ConditionVariable`. `return_from_maintenance` needs the same monitor to return connections. Neither thread can proceed:

```ruby
  # Main request thread — connection_pool.rb:1176
  def try_to_queue_for_background_connection(checkout_timeout)
    return unless @maintaining > 0

    synchronize do                  # ← acquires pool monitor
      return unless @maintaining > @available.num_waiting
      @available.poll(100)          # ← @cond.wait releases QUEUE lock, not pool monitor
    end                             #   pool monitor held for up to 100 seconds
  end

  # Maintenance thread — connection_pool.rb:993
  def return_from_maintenance(conn)
    synchronize do                  # ← needs pool monitor — BLOCKED
      conn.expire(false)
      @available.add_back(conn)
      @maintaining -= 1
    end
  end
```

The queue's `wait_poll` calls `@cond.wait(timeout)` which releases `@lock` (the queue's internal `Monitor`) but the pool's `MonitorMixin` monitor, acquired by the enclosing `synchronize`, is never released. The maintenance thread cannot return the connection it's holding, and the main thread cannot get a connection. The deadlock resolves only when `@cond.wait` hits the hardcoded `100`-second timeout.

This is a problem for us because we've set `async_query_executor` in `config/application.rb`:
<img width="942" height="68" alt="image" src="https://github.com/user-attachments/assets/fb10d161-1fa6-4299-87be-130ac8bbf181" />
(Hi @nateberkopec! 👋)
Considering that we don't use `load_async` anywhere, I'm going to just remove the config to try and unblock us. We use Unicorn so our pool likely has 1 actual connection made, regardless of our pool size of 10. So there's a possibility removing `async_query_executor` won't actually help.

The deadlock requires all pool connections to be checked out for maintenance simultaneously. This happens in two cases:

1. `async_query_executor` is configured — sequential_maintenance fans out via `@async_executor.post`, checking out all connections in parallel. Pool is fully drained while pings are in flight.
2. `pool` (max_connections) = 1 — only one connection exists, so even serial maintenance drains the pool.

Most apps have `pool >= 2` and no `async_query_executor`, so they never hit this — serial maintenance only holds one connection at a time, leaving spares available.

I had Claude whip up a reproduction script:
https://gist.github.com/ngan/5dca5aceab77d3765106e5ae21196c7e

I'm wondering if it'd make sense to move the `poll` outside of the mutex. In Puma, there's a possibility that 2 threads wait, at least only 1 will wait the whole timeeout instead of both deadlocking each other.
```ruby
  def try_to_queue_for_background_connection(checkout_timeout)
    return unless @maintaining > 0

    synchronize do
      return unless @maintaining > @available.num_waiting
    end

    begin
      @available.poll(checkout_timeout)
    rescue ConnectionTimeoutError
      nil
    end
  end
```

Again, I _think_ this is the problem. Could be completely wrong 😅 
As always, thank you for all you do for the community!

matthewd commented:
Hey @ngan, thanks for reporting this!

So far I can't work out how it could happen :confused: -- in the real implementation (and differing from your repro script) the pool and its `@available` queue share the same lock, so they can't deadlock as described.

https://github.com/rails/rails/blob/27fb2a9192b2492791528fc7c3afb53736696bc5/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L294

https://github.com/rails/rails/blob/27fb2a9192b2492791528fc7c3afb53736696bc5/activerecord/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb#L13-L14

ngan commented:
Thanks for looking into it @matthewd! We'll keep digging on our end.
